src/Simulation/Simulators/QuantumSimulator/ApplyControlledZ.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Quantum.Simulation.Core;
+
+namespace Microsoft.Quantum.Simulation.Simulators
+{
+    public partial class QuantumSimulator
+    {
+        public class QSimApplyControlledZ : Intrinsic.ApplyControlledZ
```

Comment:
We are decomposing at the Q# level as much as possible right now, and that's part of what made these necessary. Because the `Controlled` specialization allows for any number of controls, I need to have a callable that is specifically the singly controlled gate and no `Controlled` specialization. Otherwise it won't match what the target architecture actually supports. It makes sense from a pure instruction translation standpoint, but it looks very odd and redundant when layered on top of our simulator that supports more than this.

Reply:
To answer that a slightly different way, the Type1 package captures targeting a gate set is that is significantly more constrained than the one exposed by our simulator. As a result, implementing these limited intrinsics on top of our simulator requires using these C# classes as a shim that exposes less functionality. It's essentially the opposite of optimization, and it shows: simulating certain operations using this Type1 package are noticeably slower because it will use a complex decomposition, possibly including additional qubits and entanglers, instead of just calling directly into an optimized simulator API.